### ISSUE TITLE
Quarto render with --execute-params only updates executable code but not shortcode

### ISSUE THREAD

**Issue by joanbadia** [quarto-dev/quarto-cli#11167]
### Bug description

Given a main qmd document `test.qmd`, when specifying the params externally in a `params.yml` file and trying the `quarto render test.qmd --execute-params params.yaml` the parameters are only updated in the portion of executable code (e.g. those included within an R chunk), but not for the shortcodes. Is this an intentional behaviour? 

### Steps to reproduce

`test.qmd`
````qmd

---
title: "Conditional content with params"
format: html 
editor: visual
params:
  print: FALSE
---

::: {.content-hidden unless-meta="params.print"}
If PRINT parameter is TRUE, we'll show this.

shortcode shows PRINT = {{< meta params.print >}}

```{r, echo=FALSE}
cat("executed code shows PRINT = ", params$print)
if(params$print){
  cat("Hello World")
} else {
  cat("Not printing anything, because params$print is FALSE")
}
```
:::

::: {.content-hidden when-meta="params.print"}
If PRINT parameter is FALSE, we'll show this instead.

shortcode shows PRINT = {{< meta params.print >}}

```{r, echo=FALSE}
cat("executed code shows PRINT = ", params$print)
if(params$print){
  cat("Hello World")
} else {
  cat("Not printing anything, because params$print is FALSE")
}
```
:::

````

`params.yml`
````yml
print: FALSE
````

#### Conditional content with params

If PRINT parameter is FALSE, we’ll show this instead.

shortcode shows PRINT = false

executed code shows PRINT =  FALSE

Not printing anything, because params$print is FALSE

#### Conditional content with params

If PRINT parameter is FALSE, we'll show this.

shortcode shows PRINT = true

executed code shows PRINT =  FALSE

Not printing anything, because params$print is FALSE

### Your environment

- IDE: Rstudio 2024.04.2+764
- OS: Windows 10 22H2

### Quarto check output

Quarto 1.4.555
[>] Checking versions of quarto binary dependencies...
      Pandoc version 3.1.11: OK
      Dart Sass version 1.69.5: OK
      Deno version 1.37.2: OK
[>] Checking versions of quarto dependencies......OK
[>] Checking Quarto installation......OK
      Version: 1.4.555
      CodePage: 1252

[>] Checking tools....................OK
      TinyTeX: (not installed)
      Chromium: (not installed)

[>] Checking LaTeX....................OK
      Tex:  (not detected)

[>] Checking basic markdown render....OK

[>] Checking Python 3 installation....OK
      Version: 3.11.2
      Jupyter: 5.2.0
      Kernels: python3

[>] Checking Jupyter engine render....OK

[>] Checking R installation...........OK
      Version: 4.4.1
      knitr: 1.48
      rmarkdown: 2.28

[>] Checking Knitr engine render......OK

**Comment by cscheid** [comment]
Yes, `params` doesn't currently get forwarded to document metadata. The behavior _is_ intentional (if counterintuitive). The rationale is that since there are other ways to specify parameters outside of metadata, we have two options:

- reify parameters communicated outside of metadata into a "params" metadata field. This can cause confusion if users expect parameters to _not_ be merged into the metadata
- do not reify such parameters. That will cause the confusion you're expressing.

Quarto currently chooses the latter.

**Comment by mcanouil** [comment]
I don't think that is a bug.
Meta is to retrieve document (static) metadata, so it returns the correct value.

If you want to output parameters (which are "dynamic metadata"), then you need R inline code in your case `` `{r} params[["print"]]` ``.

**Comment by mcanouil** [comment]
Note that you are mixing code cells syntax, which isn't recommended by `knitr`.
Quarto uses and recommends YAML syntax.
A code cell thus should have nothing between the brackets beside the language.

````
```{r}
#| echo: false
...
````

If your titles are conditional on the parameters, why not put them in the code cells?

````
```{r}
#| output: asis
cat("\n\n## My second level header\n\n)
```
````

**Comment by joanbadia** [comment]
OK, updated the example to remove the {r, echo=FALSE}.

Why I don't put them in the code cells? Because when I have several methods inside a module, then I lose all navigation capabilities within the document when I include it within a R chunk, while the shortcode retains the navigation capabilities. Shortcodes are basically that good. Also, the titles are both for the person that will read the report and for me that might need to go back to that long module, review the code and add more.

Do you know any way to achieve the same thing as your example code with a custom shortcode? Specifically this:

```{r}
#| output: asis
cat("\n\n## My second level header\n\n)
```

Is it possible at all? I'll try to code it myself if it's possible...

**Comment by mcanouil** [comment]
@joanbadia Could you open a separate GitHub Dicussion (Q&A) to avoid cluttering this issue?

**Comment by joanbadia** [comment]
Ok, done, I'll paste the link here in case anyone stumbles into this post and also wants something similar. 

For summary, now I am trying to attempt to achieve what I want with a custom lua filter:

https://github.com/quarto-dev/quarto-cli/discussions/11173

**Comment by joanbadia** [comment]
I found a long-winded workaround that works for me because I planned to use includes anyway, see my answer in: https://github.com/quarto-dev/quarto-cli/discussions/11173

Params can be dynamically updated when you specify the following in the YAML header: 

````
---
params:
  print: "`r params$print`"
---
````

So, the workaround is defining default params in a master document and pass the `r params$...` to the child documents and render all via includes. When doing this and if you have a params.yml and use quarto render ... --execute-params params.yml, it does update and behave as I wanted. Additionally, you have some default params which never hurts.

For me, this is more than enough, as I mainly wanted reusable child qmd components.

**Comment by fkohrt** [comment]
For reference, I've been stumbling across this now as I wanted to show R code to readers which contains the citation of a package (but doesn't the reader require to install that package):

``````qmd
---
title: "Test 1"
engine: knitr
params:
  palmerpenguins_source: !expr format(citation("palmerpenguins", auto = TRUE), bibtex = FALSE, style = "text")
---

```{.r .code-overflow-wrap filename="custom.R"}
dat_source <- "{{< meta params.palmerpenguins_source >}}"
```
``````

Instead, I'm now doing the following:

``````qmd
---
title: "Test 2"
engine: knitr
---

```{r, class.output = "r code-overflow-wrap", echo = FALSE}
dat_source <- citation("palmerpenguins", auto = TRUE) |>
  format(bibtex = FALSE, style = "text")
dat_source |>
  gsub("\\n", " ", x = _) |>
  gsub("\"", "\\\\\"", x = _) |> # Escape any quotation marks
  paste0("dat_source <- \"", ... = _, "\"") |>
  cat()
```
``````

Unfortunately, I'm loosing the possibility to specify a filename with this solution.

**Comment by cderv** [comment]
> Unfortunately, I'm loosing the possibility to specify a filename with this solution. 

Maybe I'll manage to improve the params to meta idea using Quarto R package. But in the meantime you can leverage other knitr option like `attr-output`

````markdown

---
title: "Test 2"
engine: knitr
---

```{r}
#| class-output: "r code-overflow-wrap"
#| echo: false
#| attr-output: 'filename="custom.R"'
dat_source <- citation("palmerpenguins", auto = TRUE) |>
  format(bibtex = FALSE, style = "text")
dat_source |>
  gsub("\\n", " ", x = _) |>
  gsub("\"", "\\\\\"", x = _) |> # Escape any quotation marks
  paste0("dat_source <- \"", ... = _, "\"") |>
  cat()
```
````

![Image](https://github.com/user-attachments/assets/6d9d9e51-6fd8-41fc-aae1-c0979824f138)

**Comment by fkohrt** [comment]
> But in the meantime you can leverage other knitr option like `attr-output`

That's a beautiful solution for my problem, thank you!

**Comment by Steinthal** [comment]
> Maybe I'll manage to improve the params to meta idea using Quarto R package.

Using **mcanouil**'s approach, this works:

````
---
title: "Conditional content with params"
format: html 
params:
  print: false
  pca: true
  spca: false
  volcano: true
  topheatmap: false
---

```{r}
#| echo: false
#| label: meta from params
#| output: asis
cat("\n\n---\n")
cat(sprintf("print: %s\n", params[["print"]]))
cat(sprintf("pca: %s\n", params[["pca"]]))
cat(sprintf("spca: %s\n", params[["spca"]]))
cat(sprintf("volcano: %s\n", params[["volcano"]]))
cat(sprintf("topheatmap: %s\n", params[["topheatmap"]]))
cat("\n---\n\n")
```

::: {.content-hidden unless-meta="print"}
If PRINT parameter is TRUE, we'll show this.

shortcode shows PRINT = {{< meta print >}}

```{r}
#|  echo: false
cat("executed code shows PRINT = ", params$print)
if(params$print){
  cat("Hello World")
} else {
  cat("Not printing anything, because params$print is FALSE")
}
```
:::

::: {.content-hidden when-meta="print"}
If PRINT parameter is FALSE, we'll show this instead.

shortcode shows PRINT = {{< meta print >}}

```{r}
#| echo: false
cat("executed code shows PRINT = ", params$print)
if(params$print){
  cat("Hello World")
} else {
  cat("Not printing anything, because params$print is FALSE")
}
```
:::
````

**Comment by mcanouil** [comment]
@Steinthal FYI, @cderv added an helper R function for this purpose in the R package (development version not yet released to CRAN):
- https://github.com/quarto-dev/quarto-r/pull/252

**Comment by Steinthal** [comment]
Great, I'm looking forward to the release.

**Comment by joanbadia** [comment]
I've installed Quarto R package developer and updated Quarto CLI to the 1.8.14 but I can't figure out what and how I would use the `quarto::write_yaml_metadata_block` for my use case. Care to help and I'll post the result?

**Comment by cderv** [comment]
> but I can't figure out what and how I would use the quarto::write_yaml_metadata_block for my use case

From my understanding of your example

- You want to use a parameter value as a metadata to use conditional `when-meta`
- So you need to make you R object and its value available as a metadata
- To do that, you need to write a YAML metadata block 

Until now this is exactly what was explained in all above answer and example. 

What the R package brings is easy writing of this YAML metadata block. 

So here is the example

````markdown
---
title: "Conditional content with params"
format: html 
params:
  print: false
---


```{r}
#| echo: false
#| output: asis
quarto::write_yaml_metadata_block(
  params = params
)
```


::: {.content-hidden unless-meta="params.print"}
If PRINT parameter is TRUE, we'll show this.

shortcode shows PRINT = {{< meta params.print >}}

```{r}
#| echo: false
cat("executed code shows PRINT = ", params$print)
if(params$print){
  cat("Hello World")
} else {
  cat("Not printing anything, because params$print is FALSE")
}
```
:::

::: {.content-hidden when-meta="params.print"}
If PRINT parameter is FALSE, we'll show this instead.

shortcode shows PRINT = {{< meta params.print >}}

```{r}
#| echo: false
cat("executed code shows PRINT = ", params$print)
if(params$print){
  cat("Hello World")
} else {
  cat("Not printing anything, because params$print is FALSE")
}
```
:::

````

![Image](https://github.com/user-attachments/assets/80aa9cdf-364e-4961-9bad-cb119a5e194c)

Now doing `quarto render index.qmd -P print:true` leads to 

![Image](https://github.com/user-attachments/assets/7bc3de6a-db02-4bd4-8bad-2e4b22d95281)

Hope it helps

<details><summary>What the R function does</summary>
<p>

What the function writes in intermediate markdown is 
```
---
params:
  print: true
---
```

or 
````
---
params:
  print: false
---
````

</p>
</details>

**Comment by joanbadia** [comment]
> ```` 
> ```{r}
> #| echo: false
> #| output: asis
> quarto::write_yaml_metadata_block(
>   params = params
> )
> ```
> ````


Yes, that did it! It works perfectly, and I've checked and works with both:

> `quarto render test.qmd --execute-params params.yml`

And with:

> `quarto render test.qmd -P print:false`


Thank you very much!

I would highly recommend it including in the documentation when you release this as part of the stable release.

My output now: 

> Title 2
> 
> If PRINT parameter is FALSE, we’ll show this instead.
> 
> shortcode shows PRINT = false
> 
> executed code shows PRINT =  FALSE
> Not printing anything, because params$print is FALSE

**Comment by mcanouil** [comment]
Please, don't close. Let the maintainers/developers decide.

**Comment by cderv** [comment]
@joanbadia did you look at the documentation at https://quarto-dev.github.io/quarto-r/articles/dynamic-metadata.html

Was it not clear enough? I welcome feedback to improve! You can open comment at https://github.com/quarto-dev/quarto-r
